### PR TITLE
[MRG+1] add a Talairach atlas fetcher (see http://talairach.org)

### PR DESCRIPTION
**Pull request by jeromedockes**
- [x] separate atlas levels

### CODE REVIEW

**Comment by bthirion** [comment]
LGTM. +1 for merge

**Comment by KamalakerDadi** [comment]
Could you please rebase ?
The tests failures are fixed in master now.

File: nilearn/datasets/atlas.py [hunk 1]
```diff
@@ -946,3 +946,52 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
...
+    Brain Mapp 5, 238-242, 1997.`
+
+    """
+    atlas_url = 'http://www.talairach.org/talairach.nii'
+    data_dir = _get_dataset_dir(
+        'talairach_atlas', data_dir=data_dir)
+    atlas = _fetch_files(
+        data_dir, [('talairach.nii', atlas_url, {})])[0]
+    atlas = check_niimg(atlas)
+    labels = atlas.header.extensions[0].get_content()
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: You should use get_header ```from nilearn._utils.compat import get_header```.
This way we could adapt to multiple versions of nibabel.

File: nilearn/datasets/atlas.py [hunk 2]
```diff
@@ -946,3 +946,52 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
 
     return Bunch(labels=annot_left[2],  map_left=annot_left[0],
                  map_right=annot_right[0], description=fdescr)
+
+
+def fetch_atlas_talairach(data_dir=None):
+    """Download the Talairach atlas.
+
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: Can you add .. versionadded:: 0.3.2 ?

File: nilearn/datasets/atlas.py [hunk 3]
```diff
@@ -946,3 +946,52 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
...
+    Parameters
+    ----------
+    data_dir: string, optional (default=None)
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: Perhaps string -> str

File: nilearn/datasets/atlas.py (hunk 2 above)
Comment (KamalakerDadi) [reply nilearn/datasets/atlas.py]: Reference: http://www.sphinx-doc.org/en/stable/markup/para.html

File: nilearn/datasets/atlas.py [hunk 4]
```diff
@@ -946,3 +946,52 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
 
     return Bunch(labels=annot_left[2],  map_left=annot_left[0],
                  map_right=annot_right[0], description=fdescr)
+
+
+def fetch_atlas_talairach(data_dir=None):
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: Why is that we don't have other arguments like ```resume``` and ```verbose``` ?

**Comment by KamalakerDadi** [comment]
Minor comments from my side.
Once you rebase we can check for all CIs to green. Otherwise it is almost to merge.

**Comment by KamalakerDadi** [comment]
Have you looked at the atlas visualizing it ? Is it a ROI or cortical thickness ?

**Comment by jeromedockes** [comment]
> Have you looked at the atlas visualizing it ? Is it a ROI or cortical thickness ?

tal_10

tal_xyz

**Comment by jeromedockes** [comment]
they are rois with multiple levels of description (hemisphere, lobe, gyrus, tissue type,  Brodmann area)

**Comment by KamalakerDadi** [comment]
Looks great. Waiting for CIs to be green.

**Comment by jeromedockes** [comment]
> Looks great. Waiting for CIs to be green.

thanks!

**Comment by GaelVaroquaux** [comment]
Great job everybody!

**Comment by KamalakerDadi** [comment]
>they are rois with multiple levels of description (hemisphere, lobe, gyrus, tissue type, Brodmann area)

Thinking about a bit of it. May be we should also have an option to fetch each level or all levels at once.
For instance, atlas_type='hemisphere' something like this ?

**Comment by GaelVaroquaux** [comment]
That might be a good idea : to make it easier to use. 

⁣Sent from my phone, please excuse typos and briefness​

**Comment by jeromedockes** [comment]
Yes it's a good idea. However, I don't think the separate atlases are available
to fetch - we have to get the one atlas where the levels are mixed and
disentangle them ourselves. I have a naive implementation for this but it takes
around 30 seconds. I'll try to improve it but I don't think this can be done
extremely quick; we can do it when the file is fetched and store all six atlases
in nilearn_data.

**Comment by bthirion** [comment]
On the other hand, doesn't this type of feature add to the heterogeneity across atlases (I mean API-wise) ? I'm not a big fan of this idea.

**Comment by GaelVaroquaux** [comment]
> On the other hand, doesn't this type of feature add to the
> heterogeneity across atlases (I mean API-wise) ?

I think that it should be implemented in a way that it is always
returning the same type of object, just with different atlases in it. For
the exact reason that you mention.

**Comment by jeromedockes** [comment]
> On the other hand, doesn't this type of feature add to the heterogeneity across atlases (I mean API-wise) ? I'm not a big fan of this idea.

I agree. However, having labels that are records of five strings (the hemisphere, the tissue type, etc) isn't ideal either. having the choice between five different atlases with string labels may be better (and some fetchers already have this kind of interface, for example the harvard/oxford one). another option woud be to have different fetchers for each atlas, like fetch_atlas_talairach_hemisphere, fetch_atlas_talairach_gyrus, and so on - but this adds five names to the module namespace, some of which are not very useful (I don't think many people will use the hemispheres one)

**Comment by bthirion** [comment]
You're probably right that using a kw for the type of representation expected is probably the simplest and clearest approach.

File: nilearn/datasets/atlas.py [hunk 5]
```diff
@@ -11,7 +11,7 @@
 
 from .utils import _get_dataset_dir, _fetch_files, _get_dataset_descr
 from .._utils import check_niimg
-from .._utils.compat import _basestring, get_affine
+from .._utils.compat import _basestring, get_affine, get_header
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: We just bumped nibabel to 2.0.2. I think these imports get_affine and get_header can be removed.

File: nilearn/datasets/atlas.py [hunk 6]
```diff
@@ -946,3 +946,61 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
...
+    Brain Mapp 5, 238-242, 1997.`
+    """
+    atlas_url = 'http://www.talairach.org/talairach.nii'
+    data_dir = _get_dataset_dir(
+        'talairach_atlas', data_dir=data_dir)
+    atlas = _fetch_files(
+        data_dir, [('talairach.nii', atlas_url, {})],
+        resume=resume, verbose=verbose)[0]
+    atlas = check_niimg(atlas)
+    header = get_header(atlas)
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: atlas.header should be fine now after rebase.

**Comment by GaelVaroquaux** [comment]
@jeromedockes : could you rebase on master.

Also, you have an incorrect indentation in the docstring of fetch_atlas_talairach, as revealed by circleCI.

File: nilearn/datasets/atlas.py [hunk 7]
```diff
@@ -12,7 +15,7 @@
 from .utils import _get_dataset_dir, _fetch_files, _get_dataset_descr
 from .._utils import check_niimg
 from .._utils.compat import _basestring
-from ..image import new_img_like
+from ..image import new_img_like, load_img
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: May be we don't need to import load_img. We can use check_niimg

File: nilearn/datasets/atlas.py [hunk 8]
```diff
@@ -963,3 +966,149 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
...
+    The image contains all five levels of the atlas, each encoded on 8 bits
+    (least significant octet contains the hemisphere, the next one the lobe,
+    then gyrus, tissue, and ba).
+
+    The labels json file contains
+    [['level name', ['labels', 'for', 'this', 'level' ...]], ...],
+    where the levels are in the order mentionned above.
+
+    """
+    data_dir = _get_dataset_dir('talairach_atlas', data_dir=data_dir)
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: Could you please add verbose in this function ?

File: nilearn/datasets/atlas.py [hunk 9]
```diff
@@ -963,3 +966,149 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
...
+    verbose : int
+        verbosity level (0 means no message).
+
+    Returns
+    -------
+    sklearn.datasets.base.Bunch
+        Dictionary-like object, contains:
+
+        - maps: 3D Nifti image, values are indices in the list of labels.
+        - labels: list of strings. Starts with 'Background'.
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: Small question: Are these labels sorted according to maps indices ?

File: nilearn/datasets/atlas.py [hunk 10]
```diff
@@ -963,3 +966,149 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
...
+    for pos, level in enumerate(talairach_levels):
+        if verbose:
+            print(level)
+        level_img = np.zeros(atlas_img.shape, dtype=np.int64)
+        level_labels = {'*': 0}
+        for region_nb, region in enumerate(labels[:, pos]):
+            level_labels.setdefault(region, len(level_labels))
+            level_img[atlas_img.get_data() == region_nb] = level_labels[
+                region]
+        level_img <<= 8 * pos
```
Comment (KamalakerDadi) [inline nilearn/datasets/atlas.py]: Could you please add short comment about what this code block does ?

**Comment by KamalakerDadi** [comment]
I have checked that I can be able to download atlas at each level. I just want to make sure that labels are ordered according to the indices in the atlas provided. Otherwise, +1 for merge.

**Comment by GaelVaroquaux** [comment]
> I just want to make sure that labels are ordered according to the
> indices in the atlas provided.

Have you been able to check this?

**Comment by KamalakerDadi** [comment]
>I just want to make sure that labels are ordered according to the
 indices in the atlas provided.
>Have you been able to check this?

I can load the labels and see that there are 72 for brodmann and 56 for gyrus. Its a bit hard to check because I don't have anatomical knowledge.

File: nilearn/datasets/atlas.py [hunk 11]
```diff
@@ -963,3 +966,149 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
 
     return Bunch(labels=annot_left[2],  map_left=annot_left[0],
                  map_right=annot_right[0], description=fdescr)
+
+
+def _talairach_levels():
```
Comment (GaelVaroquaux) [inline nilearn/datasets/atlas.py]: Why a function and not a constant?

**Comment by jeromedockes** [comment]
> I just want to make sure that labels are ordered according to the indices in the atlas provided

They are, by construction. I have also checked this visually by looking at the hemispheres, the lobes, the gyri and the tissue types

File: nilearn/datasets/atlas.py (hunk 9 above)
Comment (jeromedockes) [reply nilearn/datasets/atlas.py]: yes, should I repeat this here or is 'values are indices in the list of labels' enough?

**Comment by KamalakerDadi** [comment]
>I just want to make sure that labels are ordered according to the indices in the atlas provided

>They are, by construction. I have also checked this visually by looking at the hemispheres, the lobes, the gyri and the tissue types

LGTM

File: nilearn/datasets/atlas.py (hunk 9 above)
Comment (KamalakerDadi) [reply nilearn/datasets/atlas.py]: I think that should be enough.

**Comment by jeromedockes** [comment]
> I just want to make sure that labels are ordered according to the indices in the atlas provided
> They are, by construction. I have also checked this visually by looking at the hemispheres, the lobes, the gyri and the tissue types
> LGTM

added some plots if you want to have a look

**Comment by KamalakerDadi** [comment]
>added some plots if you want to have a look

Looks good. Thanks

File: nilearn/datasets/atlas.py [hunk 12]
```diff
@@ -963,3 +968,150 @@ def fetch_atlas_surf_destrieux(data_dir=None, url=None,
...
+
+    `Lancaster JL, Rainey LH, Summerlin JL, Freitas CS, Fox PT, Evans AC, Toga
+    AW, Mazziotta JC. Automated labeling of the human brain: A preliminary
+    report on the development and evaluation of a forward-transform method. Hum
+    Brain Mapp 5, 238-242, 1997.`
+    """
+    levels = _TALAIRACH_LEVELS
+    if level_name not in levels:
+        raise ValueError('"level_name" should be one of {}'.format(levels))
+    position = levels.index(level_name)
```
Comment (GaelVaroquaux) [inline nilearn/datasets/atlas.py]: I would prefer if you did not define the local variable "levels", but rather used the module global "_TALAIRACH_LEVELS". It helps following the code.

**Comment by GaelVaroquaux** [comment]
Aside my little comment on the local variable vs module global, this is good to go for me. I'll merge as soon as you change this detail.

Thanks!

**Comment by jeromedockes** [comment]
> Aside my little comment on the local variable vs module global, this is good to go for me. I'll merge as soon as you change this detail.

sorry I just saw this. should be ok now

**Comment by GaelVaroquaux** [comment]
OK, merging. Thanks!

**Comment by jeromedockes** [comment]
Great! thanks to the three of you for reviewing

**Comment by GaelVaroquaux** [comment]
Thanks for the PR! Giving value to our users!
